### ISSUE TITLE
SkyFactory 3 - mob spawning causes huge TPS issues v2

### ISSUE THREAD

**Issue by KaiKikuchi** [SpongePowered/SpongeForge#1158]
Related to https://github.com/SpongePowered/SpongeForge/issues/1156

![visualvm](https://cloud.githubusercontent.com/assets/8842403/21626963/d807aaba-d214-11e6-8add-3da1cf411104.png)

https://timings.aikar.co/v2/?id=30c9da39b59d42279e405c86aabe5943&section=lag

After fixing the setEntity method the TPS is slightly better, but still there are more methods causing TPS issues.

SpongeForge: 1.10.2-2202-5.1.0-BETA-2016

**Comment by KaiKikuchi** [comment]
@Aaron1011 Thanks!

**Comment by Slind14** [comment]
![javaw_2017-01-04_12-21-46](https://cloud.githubusercontent.com/assets/3914677/21640898/7d42748e-d278-11e6-9335-b3fe7eea5e62.png)
![javaw_2017-01-04_12-22-03](https://cloud.githubusercontent.com/assets/3914677/21640897/7d425576-d278-11e6-927b-17ba5c92512f.png)

**Comment by Aaron1011** [comment]
@KaiKikuchi @Slind14: Can you both re-run whatever profile you're using, and enable the option to show the number of method calls? Make sure to expand everything out all the way (as in @Slind14's comment) when you screenshot it.

**Comment by KaiKikuchi** [comment]
@Aaron1011 I'm sorry but we're using VisualVM and a JMX connection. That doesn't show the invocation count. Unfortunately I don't know how to make it show the invocation count or use another way other than a JMX connection. I've searched on Google without any results. I tried everything I could with VisualVM and the JMX connection. If you know a way, please let me us know.

**Comment by Aaron1011** [comment]
@KaiKikuchi: You could use the evaluation version of YourKit, which supports invocation counting.

**Comment by KaiKikuchi** [comment]
I'm sorry... unfortunately our evaluation version of YourKit expired.

**Comment by bloodmc** [comment]
@KaiKikuchi Please retest with a clean sponge config. I see you turned off chunk-gc entirely which will hurt performance. Wipe out your sponge config folder, then provide another timings report.

**Comment by alicheek** [comment]
Ive got a timings report here with your sponge timings report : https://timings.aikar.co/?id=e2eb41af474c4b92b7055a81d59b59ca&section=all

This is with pretty much default settings for sponge as you asked with about 20 players on

As you can see for the potential spawn calls are very high (biggest being count(56376325))

**Comment by alicheek** [comment]
here is an updated one with 40 players on : https://timings.aikar.co/?id=f2a979852ed54c1589f41d82e6c4d209&section=lag

**Comment by Slind14** [comment]
Sorry but we can't provide any invocation counts, because tracing freezes the server down to < 0.1 TPS and call counting is incompatible with sponge. https://gist.github.com/Slind14/dc64f39f7de75eb68661edb8a07d0f61

**Comment by Slind14** [comment]
After a few more tries I managed to get a short call counting, I hope it helps in combination with the sampling tree view from before.
https://transfer.sh/%28/XG5iL/sf3-call-counting.zip%29.tar.gz

**Comment by Slind14** [comment]
Unfortunately this is still a major issue. 

I went ahead and disabled the Tinkers Construct and Random Things spawn as they have a high method call amount according to timings. But that didn't change much.

**Comment by Tencao** [comment]
I can also confirm this is a bug, however I may be able to shed a little more information regarding why Trinkers mobs seem to lag spawning. Tinkers pulls from the world data cache to check if their island is in the spawn location, perhaps something is going on with the chunk cache to cause certain mod mobs to cause far more tick time?

**Comment by Slind14** [comment]
Well I fully disabled the Tinkers Construct mob spawn, so that shouldn't be an issue.

**Comment by Tencao** [comment]
But Tinkers mob spawning was causing severe lag for us so it might be linked

Looking at the code, they basically do normal mob spawning except they check the chunk cache

**Comment by KaiKikuchi** [comment]
I think this issue is caused even before mob spawning. This should be the code that searches where a mob can be spawned.

I think we just need a way to tweak how much the server tries to find valid mob spawn locations. I don't think this is caused by inefficient methods like the `EntityTypeRegistry#getEntity`... but just the server insanely trying to find where to spawn mobs.

**Comment by TheoKah** [comment]
The problem comes from the fact that SF3 is a void map. As long as biome is 'Void' this looks stable, but as soon as one change the biome, the algorithm that is looking for a spot to place the mob is going crazy:
Algorithm picks a block, see it cannot spawn here so automatically check for its neighbour, cannot spawn there either so check for the neighbour again and so on until it reaches an island and can spawn the mob or until all blocks around have been marked as checked.

The thing is the algorithm is called EVERY SINGLE TICK for hostile mobs ( vanilla minecraft: http://minecraft.gamepedia.com/Spawn )
Spigot has an option to change the spawn rate, but sponge does not offer such an option.

When changing this might help the TPS and would be a nice feature anyways, this might also make less mobs to spawn. (That's the difference between monsters spawn rate and passive mobs spawn rates)

TLDR: 
- quick fix for TPS's sake: sponge make mobs spawn rate changeable in config
- real fix for normal spawn rate: make so the algorithm handle void worlds better

Might worth checking the mob spawn algorithm in the end since they added more islands in vanilla.

**Comment by Tencao** [comment]
This is not exclusive to the void, I have a normal world and the issues are still very present, if anything a void world would be or should be less laggy than a normal world.

To further add, I'm running a completely separate modpack and the issues are still there.

**Comment by alicheek** [comment]
@Tencao Well if the issues are still present they are definitely not as bad as it is in the void world as we have a hermitpack , infinity lite and dw20 1.10.2 server which do not have tps issues like sf3 has.

**Comment by Tencao** [comment]
Maybe, but we've certainly been hit massively by it, especially when it comes to tinkers construct slime spawning. 

Edit - Named the wrong mod for slime spawning, corrected.

**Comment by alicheek** [comment]
Just to continue the process for trying to find the problem we have now updated to skyfactory 3.0.4 and the lag is still quite present for the mobspawning events even though we have mobs disabled in the normal world and a mob spawning world thats is only the size of a small island with 8 dark rooms for mobs to spawn in.  https://timings.aikar.co/?id=dc53c677518348679a869ae1f882942d

**Comment by adracamas** [comment]
I have a theyoy on the whole mob spawns increasing TPS.

A place to check is the routines for half-slabs. I have a theroy that the mobs are still trying to spawn on them even if there is no space for them. Seems a peculiar thing.. but there may be a way to optimise the code to avoid checking every mob for the space.. mainly if there is another block butted against the top of it.

![image](https://cloud.githubusercontent.com/assets/3192842/21967679/7f3822b8-db59-11e6-94c1-2a16d9f81e16.png)


This is an a server that is running the most recent version of SkyFactory3 (1.10 but they won't tell me the version of sponge they are using)

Hope this helps and if it does let me know.

**Comment by Slind14** [comment]
I'm wondering is anyone looking into this. We provided all the asked for information to the best of our abilities. It would be nice to see some progress on this matter.

**Comment by Aaron1011** [comment]
@adracamas: It's very possible that some of the lag is coming from SkyFactory itself. However, the timing reports that I've seen all point to a significant Sponge component, which I'll be working on addressing.

**Comment by Tencao** [comment]
Once a fix is released, I will also test it on my normal world which was also experiencing large spawning lag.

**Comment by Brend-Smits** [comment]
Looking forward to a fix as well! Keep up the good work Sponge!

**Comment by Slind14** [comment]
@Aaron1011 unfortunately this is still a major issue. Any chance it could be addressed further?

**Comment by bloodmc** [comment]
I have been working on many changes that will improve performance greatly as well as fix some bugs in both SkyFactory 3 and DW20 modpacks. Currently, my changes are being tested extensively by a few servers. If you want to test, contact me on GP discord which can be found here

https://discord.gg/jy4FQDz

Once the changes have been confirmed stable after a few weeks of testing, I will push it all to bleeding/stable-5.

**Comment by Aaron1011** [comment]
@Slind14: I'll be finishing up a refactor to the event factory tody that should greatly improve performance.

**Comment by Slind14** [comment]
Thank you @Aaron1011 looking forward to it :)

**Comment by Keaton1188** [comment]
@Aaron1011  Looking forward to it! :)

**Comment by Aaron1011** [comment]
See https://github.com/SpongePowered/SpongeAPI/pull/1486

**Comment by bloodmc** [comment]
@Slind14 @KaiKikuchi @Keaton1188 @Tencao @alicheek @adracamas @Rubbertjuh 
Please confirm SF 2172+ fixes your lag issues.

**Comment by Slind14** [comment]
Unfortunately I don't see any improvement: https://timings.aikar.co/?id=327239b0338445a99758b60bd0e3eae2 

Please keep in mind that during this timings report we had a tile ent going crazy, so please use absolute numbers for comparison.

**Comment by Rob5Underscores-zz** [comment]
Same here. Turning on mob spawning drops TPS about 6. Can provide logs if necessary.

**Comment by bloodmc** [comment]
@Slind14 You are using build 2169. Update to 2172+ which includes additional spawning fixes as well as async lighting which helps.

**Comment by Brend-Smits** [comment]
Running Build 2177 at the moment, having no issues with Mob Spawning enabled [SkyFactory 3]. Thanks for the fix =]
TPS is stable at 35MS (/sponge tps)

**Comment by Slind14** [comment]
Will report back with the latest build in about half an hour.

**Comment by alicheek** [comment]
@bloodmc i can happily say its made a big difference for mobspawning events, just the lovely mods themselves causing us quite a lot of lag now XD

Thanks guys for this implementation!

**Comment by Slind14** [comment]
Here is the timings report from the latest version. The mobSpawn self time is still high: https://timings.aikar.co/?id=c0080d0343204fbf819f17a3102e1d52

**Comment by alicheek** [comment]
@Slind14  wouldnt of said 9% was much, you have to remember beforehand it could be up to 40% or more with no mobs at all just the spawning event causing the high processing time.

**Comment by Slind14** [comment]
Not for us after optimizing spawn limits.

**Comment by Brend-Smits** [comment]
9% for hostile monsters is rather normal afaik.
I'm happy with the progress and with the improvements =) Thanks!

**Comment by bloodmc** [comment]
@Slind14 That doesn't look bad to me as its only 5-6ms. Try raising tick-rate-monster in sponge's global config. By default the spawner will run every tick for monsters, try setting it to like 5 to see if that helps.

**Comment by Slind14** [comment]
hmm, I tried creating another timings report after some more uptime, but it doesn't load and further tries didn't work either: https://timings.aikar.co/?id=cfc4b29bd89948f0aacf0706de127b90
I'll try that, thanks blood.

**Comment by Brend-Smits** [comment]
Interesting Slind, I thought I was along on that issue with the Aikar website. I have that as well from time to time. Restart fixes it though.

**Comment by Slind14** [comment]
@bloodmc changing this improved the method time by around 30%. Had to use yourkit, as timings doesn't work even after /sponge timings reset.

**Comment by bloodmc** [comment]
@Slind14 OK great, just adjust it until you find the best balance. I'll leave this ticket open for a few days in case anything blows up. Just so you are aware, Ambients/Animals/Water Creatures run spawning logic once every 400 ticks by default in vanilla but monsters run spawning logic every single tick. In latest build, I added options for each type so you can control it as you please.

**Comment by phit** [comment]
I keep seeing this crash since updating http://vpaste.net/wRhZq
I this related to the modspawn changes?

**Comment by bloodmc** [comment]
ExtraUtils is cancelling a non-cancellable event. The crash cannot be more clear.

https://github.com/rwtema/extrautilities/issues/1218

**Comment by phit** [comment]
I didn't see this crash until updating, my bad!
edit: Actually I think this might be still related to this update, the stacktrace in the linked issue points at lycanitesmobs spawning, while mine is actually crashing with vanilla and it didn't happen before, and I know people have been using the quarries long before this update and the only thing that changed on the server was Sponge..

edit2: proof that it only started happening after the update http://i.imgur.com/0GmShQV.png

**Comment by Slind14** [comment]
I can confirm this, we encountered this the first time with the TEST Build, if you remember blood ;)

This is new for us too (no modpack update), can anyone confirm? https://gist.github.com/Slind14/c973235fb9a50cc89be276fddccc8168 (since the upgrade from TEST8)
https://gist.github.com/Slind14/707db266866c246240687ea97786500a (was there with TEST8 already)

I haven't seen these crashes before the called sponge versions, but it could still be unrelated.

**Comment by Rob5Underscores-zz** [comment]
@Slind14 
I'm getting the first infrequently on automated shutdowns but I've had it prior to this build also.

**Comment by Brend-Smits** [comment]
Both crashes that you are referring to have been around for a while now. Not related to sponge and have been fixed afaik

**Comment by phit** [comment]
none of you seen the ExU2 crash? or did you all disable the quarry?

**Comment by Slind14** [comment]
@Rob5Underscores do you know if you experienced this prior to 2145? That was the build where the user lookup was slightly changed: https://github.com/SpongePowered/SpongeForge/commit/6bffa19544e7076c26490e5b35af55eceb3825bd

**Comment by Slind14** [comment]
@phit we do, I also wrote that ;)

**Comment by bloodmc** [comment]
@Slind14 It doesn't matter what update caused the ExU2 crash. The fact is, the mod is CANCELLING a non-cancellable event that has crashed pure forge servers. The dev needs to fix it, it is not our issue.

Build 2145 has nothing to do with the chickenchunks crash. It NPE's here

https://github.com/MinecraftForge/MinecraftForge/blob/1.10.x/src/main/java/net/minecraftforge/common/ForgeChunkManager.java#L711

**Comment by phit** [comment]
@bloodmc Look at the crash that [I linked](http://vpaste.net/wRhZq) and the [one from the ExU](http://vpaste.net/FRxbg) tracker, the one from me is crashing with vanilla spawns the other one isn't. My guess is ExU is changing something in forge that is now also being changed by Sponge. Possibly introduced with https://github.com/SpongePowered/SpongeCommon/commit/22a4974e68c051f31d86f14d9a66f2d3dda7386d

**Comment by bloodmc** [comment]
https://github.com/MinecraftForge/MinecraftForge/blob/1.10.x/src/main/java/net/minecraftforge/event/ForgeEventFactory.java#L207

https://github.com/MinecraftForge/MinecraftForge/blob/1.10.x/src/main/java/net/minecraftforge/event/entity/living/LivingSpawnEvent.java#L116

It does NOT matter how the event is triggered. The mod is CANCELLING a non-cancellable event when it should be using result, end of discussion.

**Comment by phit** [comment]
right, I never said it didn't but it might be changing the event with ASM, which now fails due to the changes in Sponge

**Comment by bloodmc** [comment]
Uh no that would be retarded if a mod attempted to make an event cancellable.

I got tired of stating the same thing over and over, here is what ExtraUtils 2 is doing

https://gist.github.com/bloodmc/4ee3317a5db5b058cea025c56ac40b0f#file-extrautils_fun-java-L4

The mod ASSUMES every LivingSpawnEvent is cancellable.

**Comment by liach** [comment]
Another case of "Stupid modders".

**Comment by bloodmc** [comment]
Closing ticket as it has been confirmed by multiple users to be solved.
